File: 2015/go/day03/part2.go
package main

import "strconv"

func processPart2(input string) int {
	x, y := 0, 0
	rX, rY := 0, 0
	visited := make(map[string]bool)
	visited["0,0"] = true
	santa := true

	for _, c := range input {
		if santa {

			switch c {
			case '^':
				y++
			case 'v':
				y--
			case '>':
				x++
			case '<':
				x--
			}
			visited[strconv.Itoa(x)+","+strconv.Itoa(y)] = true
		} else {
			switch c {
			case '^':
				rY++
			case 'v':
				rY--
			case '>':
				rX++
			case '<':
				rX--
			}
			visited[strconv.Itoa(rX)+","+strconv.Itoa(rY)] = true
		}
		santa = !santa
	}
	return len(visited)
}

File: 2015/deno/day03/part-2.ts
export function processPart2(input: string): number {
	const trimmed = input.trim();
	const directions = trimmed.split('');

	const visited = new Set<string>();
	let x = 0;
	let y = 0;
	let y1 = 0;
	let x1 = 0;
	let santa = true;
	visited.add(`${x},${y}`);

	for (const direction of directions) {
		if (santa) {
			switch (direction) {
				case '^':
					y++;
					break;
				case 'v':
					y--;
					break;
				case '>':
					x++;
					break;
				case '<':
					x--;
					break;
			}
			visited.add(`${x},${y}`);
			santa = !santa;
		} else if (!santa) {
			switch (direction) {
				case '^':
					y1++;
					break;
				case 'v':
					y1--;
					break;
				case '>':
					x1++;
					break;
				case '<':
					x1--;
					break;
			}
			visited.add(`${x1},${y1}`);
			santa = !santa;
		}
	}
	return visited.size;
}
